File: server/src/modules/customerDiscovery.js
import { z } from 'zod';
import { clamp, normalize, round, weightedScore, band } from '../utils/scoring.js';

/* ============================================================
   MODULE 1 — CUSTOMER DISCOVERY
   Input:  business idea / problem statement.
   Logic:  can we communicate with the consumer? If yes, how many?
   Output: interaction volume + a discovery readiness score.
   ============================================================ */

export const key = 'customerDiscovery';
export const title = 'Customer Discovery';
export const action = 'SCRUMING';

export const inputSchema = z.object({
  businessIdea: z.string().min(10, 'Describe the business idea in at least 10 characters'),
  problemStatement: z.string().min(10).optional(),
  /** Can the venture actually reach its consumers? */
  consumerCommunication: z.boolean().default(false),
  /** "How many" — only meaningful when communication is possible. */
  reachableConsumers: z.number().int().nonnegative().default(0),
  interviewsCompleted: z.number().int().nonnegative().default(0),
  weeklyInteractions: z.number().int().nonnegative().default(0),
  channels: z.array(z.string()).default([]),
});

/**
 * @param {z.infer<typeof inputSchema>} input
 * @returns {{ output: object, score: number }}
 */
export function run(input) {
  const data = inputSchema.parse(input);

  // The gate: without a path to the consumer, discovery cannot proceed and the
  // "how many" question is moot.
  if (!data.consumerCommunication) {
    return {
      score: 0,
      output: {
        communicationFeasible: false,
        reachableConsumers: 0,
        interactionVolume: { weekly: 0, total: 0 },
        readiness: 'BLOCKED',
        summary:
          'No viable channel to the consumer was identified. Establish a communication path before proceeding — every downstream module depends on it.',
        recommendations: [
          'Identify at least one direct channel to the target consumer.',
          'Run 5 exploratory conversations to confirm the problem exists.',
        ],
      },
    };
  }

  // Volume sub-scores. Thresholds encode "what good looks like" at discovery stage.
  const reachScore = normalize(data.reachableConsumers, 0, 500);
  const interviewScore = normalize(data.interviewsCompleted, 0, 30);
  const cadenceScore = normalize(data.weeklyInteractions, 0, 20);
  const channelScore = normalize(data.channels.length, 0, 3);

  const score = weightedScore([
    { value: reachScore, weight: 0.3 },
    { value: interviewScore, weight: 0.35 },
    { value: cadenceScore, weight: 0.25 },
    { value: channelScore, weight: 0.1 },
  ]);

  return {
    score,
    output: {
      communicationFeasible: true,
      reachableConsumers: data.reachableConsumers,
      interactionVolume: {
        weekly: data.weeklyInteractions,
        total: data.interviewsCompleted,
        annualisedProjection: round(data.weeklyInteractions * 52),
      },
      channels: data.channels,
      breakdown: { reachScore, interviewScore, cadenceScore, channelScore },
      readiness: band(score),
      summary:
        `Consumer communication is feasible across ${data.channels.length || 'unspecified'} channel(s), ` +
        `reaching ${data.reachableConsumers} consumers with ${data.weeklyInteractions} interactions per week.`,
      recommendations: buildRecommendations(data, { interviewScore, cadenceScore }),
    },
  };
}

function buildRecommendations(data, { interviewScore, cadenceScore }) {
  const out = [];
  if (interviewScore < 50) out.push('Complete at least 15 discovery interviews before sizing the market.');
  if (cadenceScore < 50) out.push('Raise weekly consumer contact — a thin cadence produces stale signal.');
  if (data.channels.length < 2) out.push('Add a second reach channel to reduce single-channel dependency.');
  if (!out.length) out.push('Discovery volume is sufficient — proceed to sector profiling.');
  return out;
}

export default { key, title, action, inputSchema, run };

File: server_python/modules/customer_discovery.py
"""MODULE 1 — CUSTOMER DISCOVERY (port of server/src/modules/customerDiscovery.js)
Input:  business idea / problem statement.
Logic:  can we communicate with the consumer? If yes, how many?
Output: interaction volume + a discovery readiness score.
"""
from typing import List, Optional
from pydantic import BaseModel, Field

from scoring import normalize, js_round, weighted_score, band

KEY = 'customerDiscovery'
TITLE = 'Customer Discovery'
ACTION = 'SCRUMING'


class InputSchema(BaseModel):
    businessIdea: str = Field(min_length=10, description='Describe the business idea in at least 10 characters')
    problemStatement: Optional[str] = Field(default=None, min_length=10)
    consumerCommunication: bool = False
    reachableConsumers: int = Field(default=0, ge=0)
    interviewsCompleted: int = Field(default=0, ge=0)
    weeklyInteractions: int = Field(default=0, ge=0)
    channels: List[str] = []


def run(raw_input, context=None):
    data = InputSchema.model_validate(raw_input or {})

    # The gate: without a path to the consumer, discovery cannot proceed and
    # the "how many" question is moot.
    if not data.consumerCommunication:
        return {
            'score': 0,
            'output': {
                'communicationFeasible': False,
                'reachableConsumers': 0,
                'interactionVolume': {'weekly': 0, 'total': 0},
                'readiness': 'BLOCKED',
                'summary':
                    'No viable channel to the consumer was identified. Establish a communication path '
                    'before proceeding — every downstream module depends on it.',
                'recommendations': [
                    'Identify at least one direct channel to the target consumer.',
                    'Run 5 exploratory conversations to confirm the problem exists.',
                ],
            },
        }

    # Volume sub-scores. Thresholds encode "what good looks like" at discovery stage.
    reach_score = normalize(data.reachableConsumers, 0, 500)
    interview_score = normalize(data.interviewsCompleted, 0, 30)
    cadence_score = normalize(data.weeklyInteractions, 0, 20)
    channel_score = normalize(len(data.channels), 0, 3)

    score = weighted_score([
        {'value': reach_score, 'weight': 0.3},
        {'value': interview_score, 'weight': 0.35},
        {'value': cadence_score, 'weight': 0.25},
        {'value': channel_score, 'weight': 0.1},
    ])

    return {
        'score': score,
        'output': {
            'communicationFeasible': True,
            'reachableConsumers': data.reachableConsumers,
            'interactionVolume': {
                'weekly': data.weeklyInteractions,
                'total': data.interviewsCompleted,
                'annualisedProjection': js_round(data.weeklyInteractions * 52),
            },
            'channels': data.channels,
            'breakdown': {
                'reachScore': reach_score,
                'interviewScore': interview_score,
                'cadenceScore': cadence_score,
                'channelScore': channel_score,
            },
            'readiness': band(score),
            'summary':
                f'Consumer communication is feasible across {len(data.channels) or "unspecified"} channel(s), '
                f'reaching {data.reachableConsumers} consumers with {data.weeklyInteractions} interactions per week.',
            'recommendations': _build_recommendations(data, interview_score, cadence_score),
        },
    }


def _build_recommendations(data, interview_score, cadence_score):
    out = []
    if interview_score < 50:
        out.append('Complete at least 15 discovery interviews before sizing the market.')
    if cadence_score < 50:
        out.append('Raise weekly consumer contact — a thin cadence produces stale signal.')
    if len(data.channels) < 2:
        out.append('Add a second reach channel to reduce single-channel dependency.')
    if not out:
        out.append('Discovery volume is sufficient — proceed to sector profiling.')
    return out
